File: 785. Is Graph Bipartite?/solution.java
class Solution {
    public boolean isBipartite(int[][] graph) {
	    // arrays for saving vertices colors
        final boolean[] colors = new boolean[graph.length];
        final boolean[] visited = new boolean[graph.length];
        for (int v = 0; v < graph.length; v++) {
            if (!visited[v]) {
                if (!dfs(graph, v, colors, visited, false)) {
                    return false;
                }
            }
        }
        return true;
    }
    
    private boolean dfs(int[][] graph, int v, boolean[] colors, boolean[] visited, boolean adjColor) {
        if (!visited[v]) {
            visited[v] = true;
            colors[v] = !adjColor; // We color the current vertex with an opposite color of its adjacent vertex.
            for (int index = 0; index < graph[v].length; index++) {
                if (!dfs(graph, graph[v][index], colors, visited, !adjColor)) {
                    return false;
                }
            }
		// if we encounter a vertex who is already colored, if that's the same color as color of the vertex that's currently being visited, then this graph is not bipartite.	
        } else if (colors[v] == adjColor) {
            return false;
        }
        
        return true;
    }
}

File: 785. Is Graph Bipartite?/solution.py
class Solution:
    def dfs(self, graph, i, color, colors):
        # print('i: {}, colors: {}'.format(i, colors))
        if colors[i] != -1:
            return colors[i] == color
        
        colors[i] = color
        for neighbor in graph[i]:
            if not self.dfs(graph, neighbor, 0 if color == 1 else 1, colors):
                return False

        return True
        
    def isBipartite(self, graph: List[List[int]]) -> bool:
        colors = [-1] * len(graph)
        for i in range(len(graph)):
            if colors[i] == -1:
                if not self.dfs(graph, i, 0, colors):
                    return False
        return True